File: backend/app/schemas/ppqr.py
"""
pPQR schemas for the welding system backend.
"""
from typing import List, Optional, Dict, Any
from datetime import datetime, date

from pydantic import BaseModel, ConfigDict, Field


# 基础信息 schemas
class PPQRBase(BaseModel):
    """Base pPQR schema."""
    title: str = Field(..., min_length=1, max_length=200, description="标题")
    ppqr_number: str = Field(..., min_length=1, max_length=50, description="pPQR编号")
    revision: str = Field(default="A", max_length=10, description="版本号")
    status: str = Field(default="draft", max_length=20, description="状态: draft, testing, completed, converted")
    
    # 模板和模块数据
    template_id: Optional[str] = Field(None, max_length=100, description="使用的模板ID")
    module_data: Optional[Dict[str, Any]] = Field(None, description="模块数据(JSON)")
    
    # 基本信息
    test_date: Optional[date] = Field(None, description="试验日期")
    test_purpose: Optional[str] = Field(None, description="试验目的")
    reference_standard: Optional[str] = Field(None, max_length=100, description="参考标准")
    welding_process: Optional[str] = Field(None, max_length=50, description="焊接工艺")
    welder_name: Optional[str] = Field(None, max_length=100, description="焊工姓名")
    project_name: Optional[str] = Field(None, max_length=100, description="项目名称")
    
    # 试验方案
    test_variables: Optional[str] = Field(None, description="试验变量")
    number_of_specimens: Optional[int] = Field(None, ge=0, description="试样数量")
    test_matrix: Optional[str] = Field(None, description="试验矩阵")
    expected_outcome: Optional[str] = Field(None, description="预期结果")
    risk_assessment: Optional[str] = Field(None, description="风险评估")
    
    # 材料信息
    base_material_spec: Optional[str] = Field(None, max_length=100, description="母材规格")
    base_material_grade: Optional[str] = Field(None, max_length=50, description="母材牌号")
    thickness: Optional[float] = Field(None, ge=0, description="厚度")
    filler_metal_spec: Optional[str] = Field(None, max_length=100, description="填充金属规格")
    filler_metal_classification: Optional[str] = Field(None, max_length=100, description="填充金属分类")
    diameter: Optional[float] = Field(None, ge=0, description="直径")
    shielding_gas: Optional[str] = Field(None, max_length=100, description="保护气体")
    batch_number: Optional[str] = Field(None, max_length=50, description="批号")
    
    # 参数对比分析
    best_group: Optional[str] = Field(None, max_length=50, description="最佳参数组")
    best_group_reason: Optional[str] = Field(None, description="最佳参数组原因")
    heat_input_analysis: Optional[str] = Field(None, description="热输入分析")
    quality_comparison: Optional[str] = Field(None, description="质量对比")
    efficiency_analysis: Optional[str] = Field(None, description="效率分析")
    recommended_parameters: Optional[str] = Field(None, description="推荐参数")
    
    # 试验评价
    test_conclusion: Optional[str] = Field(None, max_length=50, description="试验结论")
    objectives_achieved: Optional[str] = Field(None, description="目标达成情况")
    lessons_learned: Optional[str] = Field(None, description="经验教训")
    next_steps: Optional[str] = Field(None, description="下一步计划")
    convert_to_pqr: Optional[str] = Field(None, max_length=20, description="是否转换为PQR: yes/no/pending")
    target_pqr_number: Optional[str] = Field(None, max_length=50, description="目标PQR编号")
    evaluated_by: Optional[str] = Field(None, max_length=100, description="评估人")
    evaluation_date: Optional[date] = Field(None, description="评估日期")
    
    # 备注
    notes: Optional[str] = Field(None, description="备注")
    remarks: Optional[str] = Field(None, description="说明")


class PPQRCreate(PPQRBase):
    """pPQR creation schema."""
    pass


class PPQRUpdate(BaseModel):
    """pPQR update schema."""
    title: Optional[str] = Field(None, min_length=1, max_length=200, description="标题")
    ppqr_number: Optional[str] = Field(None, min_length=1, max_length=50, description="pPQR编号")
    revision: Optional[str] = Field(None, max_length=10, description="版本号")
    status: Optional[str] = Field(None, max_length=20, description="状态")
    template_id: Optional[str] = Field(None, max_length=100, description="使用的模板ID")
    module_data: Optional[Dict[str, Any]] = Field(None, description="模块数据(JSON)")
    test_date: Optional[date] = Field(None, description="试验日期")
    test_purpose: Optional[str] = Field(None, description="试验目的")
    reference_standard: Optional[str] = Field(None, max_length=100, description="参考标准")
    welding_process: Optional[str] = Field(None, max_length=50, description="焊接工艺")
    welder_name: Optional[str] = Field(None, max_length=100, description="焊工姓名")
    project_name: Optional[str] = Field(None, max_length=100, description="项目名称")
    test_variables: Optional[str] = Field(None, description="试验变量")
    number_of_specimens: Optional[int] = Field(None, ge=0, description="试样数量")
    test_matrix: Optional[str] = Field(None, description="试验矩阵")
    expected_outcome: Optional[str] = Field(None, description="预期结果")
    risk_assessment: Optional[str] = Field(None, description="风险评估")
    base_material_spec: Optional[str] = Field(None, max_length=100, description="母材规格")
    base_material_grade: Optional[str] = Field(None, max_length=50, description="母材牌号")
    thickness: Optional[float] = Field(None, ge=0, description="厚度")
    filler_metal_spec: Optional[str] = Field(None, max_length=100, description="填充金属规格")
    filler_metal_classification: Optional[str] = Field(None, max_length=100, description="填充金属分类")
    diameter: Optional[float] = Field(None, ge=0, description="直径")
    shielding_gas: Optional[str] = Field(None, max_length=100, description="保护气体")
    batch_number: Optional[str] = Field(None, max_length=50, description="批号")
    best_group: Optional[str] = Field(None, max_length=50, description="最佳参数组")
    best_group_reason: Optional[str] = Field(None, description="最佳参数组原因")
    heat_input_analysis: Optional[str] = Field(None, description="热输入分析")
    quality_comparison: Optional[str] = Field(None, description="质量对比")
    efficiency_analysis: Optional[str] = Field(None, description="效率分析")
    recommended_parameters: Optional[str] = Field(None, description="推荐参数")
    test_conclusion: Optional[str] = Field(None, max_length=50, description="试验结论")
    objectives_achieved: Optional[str] = Field(None, description="目标达成情况")
    lessons_learned: Optional[str] = Field(None, description="经验教训")
    next_steps: Optional[str] = Field(None, description="下一步计划")
    convert_to_pqr: Optional[str] = Field(None, max_length=20, description="是否转换为PQR")
    target_pqr_number: Optional[str] = Field(None, max_length=50, description="目标PQR编号")
    evaluated_by: Optional[str] = Field(None, max_length=100, description="评估人")
    evaluation_date: Optional[date] = Field(None, description="评估日期")
    notes: Optional[str] = Field(None, description="备注")
    remarks: Optional[str] = Field(None, description="说明")
    reviewed_by: Optional[int] = Field(None, description="审核人ID")
    approved_by: Optional[int] = Field(None, description="批准人ID")

    # 文档编辑模式
    document_html: Optional[str] = Field(None, description="文档HTML内容（用于文档编辑模式）")


class PPQRResponse(PPQRBase):
    """pPQR response schema."""
    id: int
    owner_id: int
    created_at: datetime
    updated_at: datetime
    created_by: Optional[int] = None
    updated_by: Optional[int] = None
    reviewed_by: Optional[int] = None
    reviewed_date: Optional[datetime] = None
    approved_by: Optional[int] = None
    approved_date: Optional[datetime] = None
    is_active: bool = True
    # 审批相关字段
    approval_instance_id: Optional[int] = None
    approval_status: Optional[str] = None
    workflow_name: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class PPQRSummary(BaseModel):
    """pPQR summary for list views."""
    id: int
    title: str
    ppqr_number: str
    revision: str
    status: str
    test_date: Optional[date] = None
    test_conclusion: Optional[str] = None
    convert_to_pqr: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    # 审批相关字段
    approval_instance_id: Optional[int] = None
    approval_status: Optional[str] = None
    workflow_name: Optional[str] = None
    can_approve: Optional[bool] = None
    can_submit_approval: Optional[bool] = None
    submitter_id: Optional[int] = None

    model_config = ConfigDict(from_attributes=True)


class PPQRListResponse(BaseModel):
    """pPQR list response with pagination."""
    items: List[PPQRSummary] = Field(..., description="pPQR列表")
    total: int = Field(..., description="总记录数")
    page: int = Field(..., description="当前页码")
    page_size: int = Field(..., description="每页记录数")
    total_pages: int = Field(..., description="总页数")

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/services/ppqr.ts
import api from './api'

// pPQR相关的TypeScript类型定义
export interface PPQRBase {
  title: string
  ppqr_number: string
  revision?: string
  status?: string
  template_id?: string
  
  // 基本信息
  test_date?: string
  test_purpose?: string
  reference_standard?: string
  welding_process?: string
  welder_name?: string
  project_name?: string
  
  // 试验方案
  test_variables?: string
  number_of_specimens?: number
  test_matrix?: string
  expected_outcome?: string
  risk_assessment?: string
  
  // 材料信息
  base_material_spec?: string
  base_material_grade?: string
  thickness?: number
  filler_metal_spec?: string
  filler_metal_classification?: string
  diameter?: number
  shielding_gas?: string
  batch_number?: string
  
  // 参数对比分析
  best_group?: string
  best_group_reason?: string
  heat_input_analysis?: string
  quality_comparison?: string
  efficiency_analysis?: string
  recommended_parameters?: string
  
  // 试验评价
  test_conclusion?: string
  objectives_achieved?: string
  lessons_learned?: string
  next_steps?: string
  convert_to_pqr?: string
  target_pqr_number?: string
  evaluated_by?: string
  evaluation_date?: string
  
  // 备注
  notes?: string
  remarks?: string
  
  // JSONB字段 - 模块化数据
  module_data?: Record<string, any>
  
  // 参数组数据（可能有多组）
  parameter_groups?: ParameterGroup[]
  
  // 关联
  related_pqr_id?: number
}

// 参数对比组
export interface ParameterGroup {
  group_number: string
  group_description?: string
  current?: number
  voltage?: number
  travel_speed?: number
  heat_input?: number
  preheat_temp?: number
  interpass_temp?: number
  wire_feed_speed?: number
  gas_flow_rate?: number
  visual_rating?: string
  mechanical_test_result?: string
  remarks?: string
}

export interface PPQRCreate extends PPQRBase {}

export interface PPQRUpdate extends Partial<PPQRBase> {
  reviewed_by?: number
  approved_by?: number
}

export interface PPQRResponse extends PPQRBase {
  id: number
  owner_id: number
  created_at: string
  updated_at: string
  created_by?: number
  updated_by?: number
  reviewed_by?: number
  reviewed_date?: string
  approved_by?: number
  approved_date?: string
}

export interface PPQRSummary {
  id: number
  title: string
  ppqr_number: string
  revision: string
  status: string
  test_date?: string
  test_conclusion?: string
  convert_to_pqr?: string
  created_at: string
  updated_at: string
  // 审批相关字段
  approval_instance_id?: number
  approval_status?: string
  workflow_name?: string
  submitter_id?: number
  can_submit_approval?: boolean
  can_approve?: boolean
}

export interface PPQRListQuery {
  page?: number
  page_size?: number
  keyword?: string
  status?: string
  reference_standard?: string
  welding_process?: string
  test_conclusion?: string
  convert_to_pqr?: string
  sort_by?: string
  sort_order?: 'asc' | 'desc'
}

export interface PPQRListResponse {
  items: PPQRSummary[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

// 参数对比数据
export interface ParameterComparisonData {
  groups: ParameterGroup[]
  comparison_chart_data: any
  best_group_analysis: any
}

// pPQR服务类
class PPQRService {
  // 获取pPQR列表
  async list(query: PPQRListQuery = {}): Promise<PPQRListResponse> {
    const params: Record<string, any> = {}

    if (query.page) params.page = query.page
    if (query.page_size) params.page_size = query.page_size
    if (query.keyword) params.keyword = query.keyword
    if (query.status) params.status = query.status
    if (query.reference_standard) params.reference_standard = query.reference_standard
    if (query.welding_process) params.welding_process = query.welding_process
    if (query.test_conclusion) params.test_conclusion = query.test_conclusion
    if (query.convert_to_pqr) params.convert_to_pqr = query.convert_to_pqr
    if (query.sort_by) params.sort_by = query.sort_by
    if (query.sort_order) params.sort_order = query.sort_order

    const response = await api.get('/ppqr/', { params })
    // API拦截器会将响应包装成 {success: true, data: ...}
    // 所以我们需要访问 response.data
    return response.data as PPQRListResponse
  }

  // 获取单个pPQR
  async get(id: number): Promise<PPQRResponse> {
    const response = await api.get(`/ppqr/${id}`)
    return response.data as PPQRResponse
  }

  // 创建pPQR
  async create(data: PPQRCreate): Promise<PPQRResponse> {
    console.log('[PPQRService] 发送创建请求:', data)
    const response = await api.post('/ppqr/', data)
    console.log('[PPQRService] 收到响应:', response)
    // API拦截器会将响应包装成 {success: true, data: ...}
    // 所以我们需要访问 response.data
    return response.data as PPQRResponse
  }

  // 更新pPQR
  async update(id: number, data: PPQRUpdate): Promise<PPQRResponse> {
    const response = await api.put(`/ppqr/${id}`, data)
    return response.data as PPQRResponse
  }

  // 删除pPQR
  async delete(id: number): Promise<void> {
    await api.delete(`/ppqr/${id}`)
  }

  // 批量删除pPQR
  async batchDelete(ids: number[]): Promise<void> {
    await api.post('/ppqr/batch-delete', { ids })
  }

  // 更新pPQR状态
  async updateStatus(id: number, status: string): Promise<PPQRResponse> {
    const response = await api.patch(`/ppqr/${id}/status`, { status })
    return response.data as PPQRResponse
  }

  // 复制pPQR
  async duplicate(id: number): Promise<PPQRResponse> {
    const response = await api.post(`/ppqr/${id}/duplicate`)
    return response.data as PPQRResponse
  }

  // 转换pPQR为PQR
  async convertToPQR(id: number, pqrData?: Partial<any>): Promise<any> {
    const response = await api.post(`/ppqr/${id}/convert-to-pqr`, pqrData || {})
    return response.data
  }

  // 获取参数对比数据
  async getParameterComparison(id: number): Promise<ParameterComparisonData> {
    const response = await api.get(`/ppqr/${id}/parameter-comparison`)
    return response.data as ParameterComparisonData
  }

  // 导出pPQR为PDF
  async exportPDF(id: number): Promise<Blob> {
    const response = await api.get(`/ppqr/${id}/export/pdf`, {
      responseType: 'blob'
    })
    return response.data as Blob
  }

  // 导出pPQR为Excel
  async exportExcel(id: number): Promise<Blob> {
    const response = await api.get(`/ppqr/${id}/export/excel`, {
      responseType: 'blob'
    })
    return response.data as Blob
  }

  // 导出参数对比报告
  async exportComparisonReport(id: number): Promise<Blob> {
    const response = await api.get(`/ppqr/${id}/export/comparison-report`, {
      responseType: 'blob'
    })
    return response.data as Blob
  }

  // 获取pPQR统计信息
  async getStatistics(): Promise<any> {
    const response = await api.get('/ppqr/statistics')
    return response.data
  }
}

// 导出单例
const ppqrService = new PPQRService()
export default ppqrService

